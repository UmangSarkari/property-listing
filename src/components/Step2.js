import React, { useState, useCallback, useMemo } from "react";
import {
  Button,
  FormControl,
  Typography,
  FormLabel,
  OutlinedInput,
} from "@material-ui/core";
import makeStyles from "@material-ui/core/styles/makeStyles";
import PropTypes from "prop-types";

import { STRINGS } from "../strings";

const styles = makeStyles((theme) => ({
  button: {
    color: "white",
    borderRadius: 16,
    width: 250,
    margin: "6px 5px",
    padding: "12px 16px",
  },
  error: {
    fontSize: "14px",
    textAlign: "center",
    color: "red",
  },
  input: {
    borderRadius: 14,
    maxWidth: 400,
    fontWeight: 500,
  },
  label: {
    margin: "6px 10px",
    fontSize: "16px",
    lineHeight: "22px",
    fontWeight: "500",
    color: "black",
  },
  step2: {
    marginTop: 20,
  },
  submit: {
    marginTop: 20,
    flexDirection: "row",
  },
}));

const FormField = ({ label, error, children }) => {
  let classes = styles();
  return (
    <FormControl fullWidth>
      <FormLabel>
        <Typography className={classes.label}>{label}</Typography>
        {children}
        {error && (
          <Typography variant="h6" className={classes.error}>
            {error}
          </Typography>
        )}
      </FormLabel>
    </FormControl>
  );
};

const Step2 = ({ data = {}, setData, setStep3 }) => {
  const [error, setError] = useState({});
  let classes = styles();

  const checkValidations = useCallback((data) => {
    let errors = {};
    if (!data?.address) {
      errors["address"] = STRINGS.ADDRESS_REQUIRED;
    } else {
      delete errors["address"];
    }
    if (!data?.bedroom || data.bedroom > 10) {
      errors["bedroom"] = STRINGS.BEDROOM_VALIDATION;
    } else {
      delete errors["bedroom"];
    }
    if (!data?.bathroom || data.bathroom > 5) {
      errors["bathroom"] = STRINGS.BATHROOM_VALIDATION;
    } else {
      delete errors["bathroom"];
    }
    if (Object.keys(errors).length) {
      setError(errors);
    } else {
      return true;
    }
  }, []);

  const submit = useCallback(() => {
    if (checkValidations(data)) {
      setStep3(true);
    }
  }, [data, setStep3]);

  const handleData = useCallback((value, key) => {
    setData((prevState) => ({
      ...prevState,
      [key]: ["bedroom", "bathroom"].includes(key) ? Number(value) : value,
    }));
  }, []);

  const isSubmitDisabled = useMemo(
    () => !data?.address || !data?.bedroom || !data?.bathroom,
    [data]
  );

  return (
    <div id="step2" className={classes.step2}>
      <FormField label={STRINGS.ADDRESS} error={error?.address}>
        <OutlinedInput
          id="address"
          type="text"
          onChange={(e) => handleData(e.target.value, "address")}
          fullWidth
          className={`${classes.input}`}
          value={data?.address}
          //   error={error?.address}
          placeholder={STRINGS.ENTER_ADDRESS}
        />
      </FormField>
      <FormField label={STRINGS.BEDROOM} error={error?.bedroom}>
        <OutlinedInput
          id="bedroom"
          type="number"
          onChange={(e) => handleData(e.target.value, "bedroom")}
          fullWidth
          className={`${classes.input}`}
          value={data?.bedroom}
          //   error={error?.bedroom}
          placeholder={STRINGS.ENTER_BEDROOMS}
        />
      </FormField>
      <FormField label={STRINGS.BATHROOM} error={error?.bathroom}>
        <OutlinedInput
          id="bathroom"
          type="number"
          onChange={(e) => handleData(e.target.value, "bathroom")}
          fullWidth
          className={`${classes.input}`}
          value={data?.bathroom}
          //   error={error?.bathroom}
          placeholder={STRINGS.ENTER_BATHROOM}
        />
      </FormField>
      <FormField label={STRINGS.DESCRIPTION} error={error?.description}>
        <OutlinedInput
          id="description"
          type="text"
          onChange={(e) => handleData(e.target.value, "description")}
          fullWidth
          className={`${classes.input}`}
          value={data?.description}
          error={error?.description}
          placeholder={STRINGS.ENTER_DESCRIPTION}
          multiline
        />
      </FormField>
      <Button
        className={`${classes.button} ${classes.submit}`}
        onClick={submit}
        color="primary"
        variant={"contained"}
        disabled={isSubmitDisabled}
      >
        Submit
      </Button>
    </div>
  );
};

Step2.propTypes = {
  data: PropTypes.object,
  setData: PropTypes.func,
  setStep3: PropTypes.func,
};

export default Step2;
