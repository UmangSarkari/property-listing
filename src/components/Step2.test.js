import { render } from "@testing-library/react";
import Step2 from "./Step2";

test("Renders Step2 Correctly", () => {
  const data = {
    address: "Sothe Africa",
    bedroom: 4,
    bathroom: 2,
    description: "Lucky you to get the optional description here.",
  };
  const setData = jest.fn();
  const setStep3 = jest.fn();
  const { asFragment } = render(
    <Step2 data={data} setData={setData} setStep3={setStep3} />
  );
  expect(asFragment()).toMatchSnapshot();
});
