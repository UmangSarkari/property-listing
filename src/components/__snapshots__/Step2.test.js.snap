// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Step2 Correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-step2-5"
    id="step2"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-filled"
      >
        <p
          class="MuiTypography-root makeStyles-label-4 MuiTypography-body1"
        >
          Address
        </p>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root makeStyles-input-3 MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="address"
            placeholder="Enter your address here"
            type="text"
            value="Sothe Africa"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-8"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </label>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-filled"
      >
        <p
          class="MuiTypography-root makeStyles-label-4 MuiTypography-body1"
        >
          Bedroom
        </p>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root makeStyles-input-3 MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="bedroom"
            placeholder="Enter number of bedrooms"
            type="number"
            value="4"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-8"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </label>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-filled"
      >
        <p
          class="MuiTypography-root makeStyles-label-4 MuiTypography-body1"
        >
          Bathroom
        </p>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root makeStyles-input-3 MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="bathroom"
            placeholder="Enter number of bathrooms"
            type="number"
            value="2"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-8"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </label>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-filled"
      >
        <p
          class="MuiTypography-root makeStyles-label-4 MuiTypography-body1"
        >
          Description (Optional)
        </p>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root makeStyles-input-3 MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
            id="description"
            placeholder="Enter description"
            rows="1"
            style="height: 0px; overflow: hidden;"
          >
            Lucky you to get the optional description here.
          </textarea>
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-8"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </label>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-1 makeStyles-submit-6 MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
